### PR TITLE
Add test for result as a subproject of another, larger, project

### PR DESCRIPTION
**Pull request by DimanNe**

### CODE REVIEW

**Comment by DimanNe** [comment]
@friendlyanon Just wondering... Do you know how I can fix the issue with `INSTALL(...)` trying to install other targets:
```
CMake Error: install(EXPORT "resultTargets" ...) includes target "result" which requires target "cxxabi" that is not in any export set.
```
As I told, I am using a monorepo in which top-level CMakeLists.txt sets/determines which low-level libraries should be used by all subprojects/submodules (one of such low-level libraries is libc++) as well as defines global sanitiser flags and so on... This is the only option if you want to use sanitisers (msan in particular). The problem is that if there is a subproject that uses `INSTALL()`, I get the error message above.

In this PR I created a [minimal reproduction example](https://github.com/DimanNe/result/pull/5/commits/d61f3574a6efc7efffa8926b9d2eecca15b9d5f3).

**Comment by friendlyanon** [comment]
The way I left things allowed the library to be [trivially consumed](https://github.com/DimanNe/result/blob/f504b5a6f0bfa68c3359ea6b6c6fedafaa7e2cac/test/CMakeLists.txt#L14-L18), I don't know what's happening in this PR.  
The project was already being consumed as a subproject by the test lists file.

You are adding a dependency using `add_subdirectory()` which is missing [`cmake_minimum_required`](https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html) to set up policies, [`project`](https://cmake.org/cmake/help/latest/command/project.html) to set up variables, you are also using `link_libraries` which is not modern CMake.

You're also adding `.cpp` source files and marking your library as static despite having totally empty TUs. I also don't understand this.

If you rolled back to f504b5a6f0bfa68c3359ea6b6c6fedafaa7e2cac and explained what it is that you want to do, then I can make another PR and show you, because as things are now this library cannot be trivially consumed with `FetchContent` by users.

**Comment by DimanNe** [comment]
Sorry for the confusion, let me explain :)

> The project was already being consumed as a subproject by the test lists file.

Yes. The problem, however, is that tests does not try to add some low-level libraries to all subprojects via this:
```
# Enable building/linking with libc++ for all projects below
add_subdirectory(subproject_test/cxxabi)
link_libraries(cxxabi)
```
(in reality I am trying to add `cxx_static` `cxxabi_static` from llvm-project)

> you are also using `link_libraries` which is not modern CMake

Yep, I suppose you are saying that projects should be using **target**_link_libraries. The problem is that there are low-level libraries (std::) and compiler flags (regarding sanitisers, `-rdynamic` for [backtraces](https://stackoverflow.com/a/37433934/758986), `-g` for debug info), that the libraries (as far as I understand) should not be adding themselves. Instead, they should be added from the most top-level CMakeLists.txt.

> You're also adding .cpp source files and marking your library as static despite having totally empty TUs. I also don't understand this.

So, here are a couple of reasons:
1. Regardless of whether TUs are empty or not, I am inclined to say that if a target consists of files (any files, it might be just headers, or .proto files or any other), they should be listed.
2. Unfortunately, I cannot find where, but it is recommended to include headers of header-only libraries in corresponding cpp's. I think reasoning is this: (1) compiler in this case can perform some very basic checking and (2) should you need to extend the library with a "normal" function / any kind of helper, everything will be in place (it will be cheap).
3. Very pragmatic reason: I suppose it is a consequence of the 1st reason: QtCreator does not show the files if they "are not part of any project" and therefore it does not understand includes, it cannot apply formatting and so on...

> If you rolled back to f504b5a and explained what it is that you want to do

Will do!

**Comment by DimanNe** [comment]
Created https://github.com/DimanNe/result/pull/6 as a demo

**Comment by friendlyanon** [comment]
Libraries have no business modifying flags they don't absolutely need. If you read the documentation, you have a lot of control over dependencies you consume and you can [inject flags](https://cmake.org/cmake/help/latest/command/project.html#code-injection) without tainting the library's lists file, however your library does not have any build artifacts so it really does have no business with flags other than what is necessary (use libc++ on Clang and others on MSVC) for the very reason that those are not the library's usage requirements and header only libraries do not have build requirements (because they don't build anything).

`INTERFACE` libraries without sources is [fixed in CMake 3.19](https://cmake.org/cmake/help/latest/command/add_library.html#interface-libraries) which will be released very soon.  
There is no clean solution before that that wouldn't needlessly litter the lists file.

Edit: I guess going with the above mentioned you can for the time being edit the lists file in `all` as it was intended to be loaded by IDEs anyway:
```cmake
project(BuildAll CXX)

set(ide_target_fix_path "${PROJECT_BINARY_DIR}/ideTargetFix.cmake")
file(WRITE "${ide_target_fix_path}" [=[
# Generated file DO NOT EDIT

set(ide_target_fix_targets "result")
set(ide_target_fix_result_sources
        include/result/coresult.h
        include/result/coresult_with_using.h
        include/result/result.h
        include/result/result_with_using.h)

macro(add_library TARGET)
  if(${TARGET} IN_LIST ide_target_fix_targets)
    add_custom_target(ide_fix_${TARGET} SOURCES ${ide_target_fix_${TARGET}_sources})
  endif()

  # Call to original add_library
  _add_library(${ARGV})
endmacro()
]=])

set(CMAKE_PROJECT_result_INCLUDE_BEFORE "${ide_target_fix_path}" CACHE INTERNAL "")

add_subdirectory("${CMAKE_CURRENT_LIST_DIR}/../test" "${CMAKE_BINARY_DIR}/all/test")
```

**Comment by DimanNe** [comment]
Can you please elaborate more on the first part of your message?
So, imagine I want to inject [libcxx](https://github.com/llvm/llvm-project/tree/master/libcxx) (emulate `link_libraries(cxx_static)` in the root CMakeLists.txt).

Am I right I need to:
1. make libcxx available (by `add_subdirectory` or by `FetchContent`?) from the root CMakeLists.txt
2. figure out path to the static library (built by `libcxx`) and path of includes of `libcxx` (can it be done in cmake by inspecting some properties of target `cxx_static`?).
3. Write `inject.cmake` file that would append the path to library path to linker flags and append include path to compiler flags.
4. Set `CMAKE_PROJECT_INCLUDE_BEFORE` to point to the `inject.cmake` and from now on, when I include any subproject (via add_subdirectory or FetchContent) the two flags will be prepended to all targets/libraries built by the subproject.

Is that right?

**Comment by friendlyanon** [comment]
libc++ was already declared as a usage requirement: https://github.com/DimanNe/result/blob/f504b5a6f0bfa68c3359ea6b6c6fedafaa7e2cac/CMakeLists.txt#L29-L31

The example with injection was w.r.t IDE integration until CMake 3.19 is released.

**Comment by DimanNe** [comment]
> libc++ was already declared as a usage requirement

It was declared as dependency only because this particular project uses some feature (coroutines) that happened to be implemented only in libc++. Imagine it were implemented in all compilers, we would remove the line from **that** CMakeLists.txt.

Ok. let me rephrase my question. Imagine there is target `asdf` (that consists of a static library and include path), which I have to make a dependency of each and every subproject I build. How do I do it in modern CMake?

**Comment by friendlyanon** [comment]
You explicitly declare and use the project that provides that target.
```cmake
# If this was declared already then the command is a noop
FetchContent_Declare(project_with_asdf ...)
FetchContent_MakeAvailable(project_with_asdf)

target_link_libraries(my_target PRIVATE project::asdf)
```
This makes everything visible and there is no implicit hidden magic going on that will just make things difficult.

**Comment by DimanNe** [comment]
It looks like there is some misunderstanding :)

What if there are many libraries in my root project (hundreds) and I would like to specify very low-level flags/libraries? For example, I would like all code in my monorepo to use  `tcmalloc` memory allocation library, or be built with some specific sanitiser flags (for `msan` to work, all code *must* be built with it - there is no other option). In this case, it seems, it is infeasible for me to modify each of the libraries by adding `target_link_libraries(my_target PRIVATE project::asdf)`, right?

My understanding of cmake was this:
* By default in most cases everyone should be using **target**_link_libraries.
* But for very-low level and project-wide settings, we should be using `link_libraries`.

And this works. It works as long as there is no INSTALL() targets...

I understand your concerns, but I do not see how I can accomplish this without using old / ugly / implicit `link_libraries`...

**Comment by friendlyanon** [comment]
For such modifications you have [toolchain files](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html) that affect your build, which should be nowhere in your lists files.

CMake does have all the knobs and switches you need, you're just trying to turn the wrong ones.

**Comment by DimanNe** [comment]
Ok. I think I have already spent too much of your time, so feel free to ignore the comment :)

Bun in case you are interested...
`clang-toolchain-file` looks promising indeed, but it seems I cannot invoke compilation of `libcxx` from within it. I tried to write this file:
```
cmake_minimum_required(VERSION 3.13)

set(CMAKE_SYSTEM_NAME Linux)
set(CMAKE_MAKE_PROGRAM make)

##############################################################################################################
# 2. libc++ / libc++abi
##############################################################################################################
add_definitions(-Wno-everything) # Ignore all 3rd-party warnings

set(LLVM_MAIN_SRC_DIR ${PROJECT_SOURCE_DIR}/contrib/llvm-project/libcxx)
set(LIBCXX_CONFIGURE_IDE)
set(LIBCXX_CXX_ABI default)
set(LIBCXX_INCLUDE_BENCHMARKS OFF)

add_subdirectory(contrib/llvm-project/libcxx)
set(LIBCXXABI_LIBCXX_INCLUDES ${PROJECT_SOURCE_DIR}/contrib/llvm-project/libcxx/include)

add_subdirectory(contrib/llvm-project/libcxxabi)
remove_definitions(-Wno-everything)
```

But I am getting this:
```
Running /CMake/bin/cmake -S /home/dimanne/devel/scripts -B /home/dimanne/devel/build-scripts-Desktop-Debug in /home/dimanne/devel/build-scripts-Desktop-Debug.
-- Looking for fopen in c
-- Configuring incomplete, errors occurred!
CMake Error at /CMake/share/cmake-3.18/Modules/CheckLibraryExists.cmake:64 (message):
  CHECK_FUNCTION_EXISTS needs either C or CXX language enabled
Call Stack (most recent call first):
  contrib/llvm-project/libcxx/cmake/config-ix.cmake:12 (check_library_exists)
  contrib/llvm-project/libcxx/CMakeLists.txt:481 (include)
```

it looks like it is **too** low-level, and it is **not** supposed to be used for compilation of anything. cmake assumes that there **exists** a compiler and everything that it needs to work properly (in my case it is `libcxx`), which is not the case, due to msan and workflow in general. And without `add_subdirectory(contrib/llvm-project/libcxx)` and without compilation it from within the file itself, I do not have any library to append (I do not know its name, I do not know its includes and so on...).

**Comment by friendlyanon** [comment]
So you just want to use libc++ in your build.<br />
This can be solved in two ways:
<ol>
<li>Simply install libc++ to the system and use <code>-stdlib=libc++</code> flag.</li>
<li>Install it locally and set the flags manually. Example where this was done:<br />
<a href="https://github.com/David-Haim/concurrencpp/blob/develop/script/install_libcxx.sh"><code>install_libcxx.sh</code></a> to install libc++<br />
<a href="https://github.com/David-Haim/concurrencpp/blob/develop/cmake/ciBuild.cmake#L25-L29"><code>ciBuild.cmake</code></a> to set the necessary flags using variables <u>intended</u> for this use by CMake<br />
<a href="https://github.com/David-Haim/concurrencpp/blob/develop/.github/workflows/ci.yml#L67-L81"><code>ci.yml</code></a> where these are tied together</li>
</ol>

Neither of these methods are visible from the lists files.

**Comment by DimanNe** [comment]
> So you just want to use libc++ in your build

Doing just *this* would be easy, indeed. I, however, want `libc++` to be a part of my mono-repo build.
Why?

Three reasons:
1. **Sanitisers**: I really want to use sanitisers, in particular msan, for which it is required that libc++ (and everything else) be built with the sanitiser flags: "[***MemorySanitizer requires that all program code is instrumented. This also includes any libraries that the program depends on, even libc.***](https://clang.llvm.org/docs/MemorySanitizer.html#handling-external-code)" (and there is no msan-built version of `libc` in Ubuntu repos)

2. **Fuzzers**: I might be wrong, but clang's libFuzzer needs to instrument code to have a good coverage-guided fuzzing.

3. Second reason is that there is no advantages of mono-repo build for libc if it is not part of mono-repo, in particular, I will not be able to look at the source code, I will not be able to set breakpoints inside the library and so on...

But I think the 1st reason is the most important for me.

**Comment by friendlyanon** [comment]
I think the second option I listed above is perfect for your usecase.
